File: app/src/main/java/ua/hospes/nfs/marathon/data/race/storage/RaceDbStorage.java
package ua.hospes.nfs.marathon.data.race.storage;

import android.content.ContentValues;

import java.util.List;

import javax.inject.Inject;

import rx.Observable;
import ua.hospes.dbhelper.InsertResult;
import ua.hospes.dbhelper.QueryBuilder;
import ua.hospes.dbhelper.UpdateResult;
import ua.hospes.nfs.marathon.core.db.DbHelper;
import ua.hospes.nfs.marathon.core.db.tables.Race;
import ua.hospes.nfs.marathon.core.db.tables.Sessions;
import ua.hospes.nfs.marathon.data.race.mapper.RaceMapper;
import ua.hospes.nfs.marathon.data.race.models.RaceItemDb;
import ua.hospes.nfs.marathon.data.race.operations.UpdateRaceOperation;

/**
 * @author Andrew Khloponin
 */
public class RaceDbStorage {
    private final DbHelper dbHelper;

    @Inject
    public RaceDbStorage(DbHelper dbHelper) {
        this.dbHelper = dbHelper;
    }


    public Observable<InsertResult<RaceItemDb>> add(List<RaceItemDb> items) {
        return dbHelper.insert(Race.name, items);
    }

    public Observable<UpdateResult<RaceItemDb>> update(RaceItemDb item) {
        return dbHelper.update(new QueryBuilder(Race.name).where(Race._ID + " = ?", String.valueOf(item.getId())), item);
    }

    public Observable<Integer> remove(RaceItemDb session) {
        return dbHelper.delete(new QueryBuilder(Race.name).where(Race._ID + " = ?", String.valueOf(session.getId())));
    }

    public Observable<Boolean> updateRaces(Iterable<UpdateRaceOperation> operations) {
        return dbHelper.multiOperationTransaction(operations);
    }


    public Observable<RaceItemDb> get() {
        return dbHelper.singleQuery(RaceMapper::map, new QueryBuilder(Race.name));
    }

    public Observable<List<RaceItemDb>> listen() {
        return dbHelper.query(RaceMapper::map, new QueryBuilder(Race.name), Race.name, Sessions.name);
    }

    public Observable<Void> reset() {
        ContentValues cv = new ContentValues();
        cv.put(Race.TEAM_NUMBER, -1);
        cv.put(Race.SESSION_ID, -1);
        return dbHelper.update(new QueryBuilder(Race.name), cv).map(result -> null);
    }

    public Observable<Void> clean() {
        return dbHelper.delete(new QueryBuilder(Race.name)).map(integer -> null);
    }
}

File: app/src/main/java/ua/hospes/rtm/core/db/race/storage/RaceDbStorage.kt
package ua.hospes.rtm.core.db.race.storage

import android.content.ContentValues
import android.database.sqlite.SQLiteDatabase
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.functions.Function
import ua.hospes.dbhelper.InsertResult
import ua.hospes.dbhelper.UpdateResult
import ua.hospes.dbhelper.builder.DeleteQuery
import ua.hospes.dbhelper.builder.SelectQuery
import ua.hospes.dbhelper.builder.UpdateQuery
import ua.hospes.dbhelper.builder.conditions.Condition
import ua.hospes.rtm.core.db.DbHelper
import ua.hospes.rtm.core.db.tables.Race
import ua.hospes.rtm.core.db.tables.Sessions
import ua.hospes.rtm.core.db.race.mapper.RaceMapper
import ua.hospes.rtm.core.db.race.models.RaceItemDb
import ua.hospes.rtm.core.db.race.operations.UpdateRaceOperation
import javax.inject.Inject

class RaceDbStorage @Inject
internal constructor(private val dbHelper: DbHelper) {

    fun add(items: List<RaceItemDb>): Observable<InsertResult<RaceItemDb>> = dbHelper.insert(Race.name, SQLiteDatabase.CONFLICT_ABORT, items)

    fun update(item: RaceItemDb): Observable<UpdateResult<RaceItemDb>> =
            dbHelper.update(UpdateQuery(Race.name).where(Condition.eq(Race.ID, item.id)), item)

    fun updateRaces(operations: Iterable<UpdateRaceOperation>): Observable<Boolean> = dbHelper.multiOperationTransaction(operations)


    fun get(): Observable<RaceItemDb> = dbHelper.querySingle({ RaceMapper.map(it) }, SelectQuery(Race.name))

    fun listen(): Observable<List<RaceItemDb>> = dbHelper.query(Function { RaceMapper.map(it) }, SelectQuery(Race.name), Race.name, Sessions.name)

    fun listen(id: Int): Observable<List<RaceItemDb>> =
            dbHelper.query(Function { RaceMapper.map(it) }, SelectQuery(Race.name).where(Condition.eq(Race.ID, id)), Race.name, Sessions.name)

    fun reset(): Observable<Void> {
        val cv = ContentValues()
        cv.put(Race.TEAM_NUMBER.name(), -1)
        cv.put(Race.SESSION_ID.name(), -1)
        return dbHelper.updateCV(UpdateQuery(Race.name), cv).map { null }
    }


    fun remove(session: RaceItemDb): Single<Int> = dbHelper.delete(DeleteQuery(Race.name).where(Condition.eq(Race.ID, session.id)))

    fun removeAll(): Single<Int> = dbHelper.delete(DeleteQuery(Race.name))
}